designs/NuGet-ClientCertificates.md

```diff
@@ -0,0 +1,114 @@
+
+It is necessary to add client certificate authentication feature support to NuGet.
+
+## Who are the customers
+
+All .NET Core customers.
+
+## Requirements
+
+* Ability to specify client certificates from common [certificate stores](https://docs.microsoft.com/en-us/dotnet/framework/wcf/feature-details/working-with-certificates#certificate-stores)
+* Ability to specify client certificates from external files in standard [formats](https://en.wikipedia.org/wiki/X.509#Certificate_filename_extensions)
+* Ability to specify inline client certificates in Base64 encoded DER certificate format
+
+## Solution
+
+1) Find specified in NuGet configuration file certificates (common implementation for [all products](https://github.com/NuGet/NuGet.Client/tree/dev/src/NuGet.Core/NuGet.Configuration) )
+2) Apply them prior any web request to appropriate http handler (product specific)
+
+## NuGet configuration changes
+
+Added new configuration section **clientCertificates** which may have children of 2 types:
+
+**fromStorage** - item for certificate import from [certificate store](https://docs.microsoft.com/en-us/dotnet/framework/wcf/feature-details/working-with-certificates#certificate-stores). Internally uses [X509Certificate2Collection.Find](https://docs.microsoft.com/en-us/dotnet/api/system.security.cryptography.x509certificates.x509certificate2collection.find?view=netframework-4.8#System_Security_Cryptography_X509Certificates_X509Certificate2Collection_Find_System_Security_Cryptography_X509Certificates_X509FindType_System_Object_System_Boolean_) method.
+Can be configured with 4 Add item's
+
+- Optional Key="**StoreLocation**" Value="[possible values](https://docs.microsoft.com/en-us/dotnet/api/system.security.cryptography.x509certificates.storelocation?view=netframework-4.8#fields)".  
+Equals '**CurrentUser**' by default
+- Optional Key="**StoreName**" Value="[possible values](https://docs.microsoft.com/en-us/dotnet/api/system.security.cryptography.x509certificates.storename?view=netframework-4.8#fields)". 
+Equals '**My**' by default
+- Optional Key="**FindType**" Value="[possible values](https://docs.microsoft.com/en-us/dotnet/api/system.security.cryptography.x509certificates.x509findtype?view=netframework-4.8#fields)". 
+Equals '**FindByThumbprint**' by default
+- Optional Key="**FindValue**" Value="The search criteria as an object"
+
+- **fromCert** - item for certificate import directly from file (DER encoded x.509, Base-64 encoded x.509, PKCS)
+Can be configured with 2 Add item's and item text body
+
+- Key="**Path**" Value="Absolute or relative path to certificate file". Relative path resolve have 2 stages: first relative to configuration file origin, second relative to current application directory.
+- Optional Key="**Password**" Value="Encrypted password". Password for certificate file. Encrypted in same manner as PackageSourceCredential password.
```

Comment:
PackageSourceCrendnetial is encrypted on Windows using the ProtectedData API [which is, only will only ever be Windows-only](https://github.com/dotnet/corefx/issues/22510#issuecomment-317197092). So, password will need a plain-text version as well, otherwise using certificaes with passwords becomes Windows-only.

Reply:
Oh I didn't realize this. Maybe you have alternative solution for NuGet components? I do not believe that plain text password is best solution :)

Reply:
Anyway ability to store password as plain text implemented